**Ponte a prueba 1**

Ahora vamos a poner a prueba tus conocimientos con este ejercicio.

Realiza una red neuronal con los siguientes requerimientos:
- Tres inputs de 10, 30 y 40.
- Un bias de 1.
- Dos capas ocultas.
- Analiza los resultados.

In [ ]:
#Respuesta con comentarios para explicar el proceso

In [ ]:
## Partimos de las funciones definidas en el notebook de la semana 17

import numpy as np

# Función de activación: sigmoide
def sigmoide(x):
    return 1 / (1 + np.exp(-x))

# Definición de la clase Neurona
class Neurona:
    def __init__(self, pesos, bias):
        self.pesos = pesos
        self.bias = bias

    def activacion(self, inputs):
        total = np.dot(self.pesos, inputs) + self.bias
        return sigmoide(total)


Ahora vamos a modificar la clase MiRedNeuronal para que se ajuste a los requerimientos:

- Tendremos tres entradas, por lo que los pesos de cada neurona en la primera capa oculta serán un array de 3 elementos.
- Habrá una segunda capa oculta con dos neuronas.
- Cada neurona en la capa de salida se alimentará con las salidas de las neuronas de la segunda capa oculta.
- El bias será de 1 para todas las neuronas.

In [ ]:
class MiRedNeuronal:
    def __init__(self):
        # Pesos y bias iniciales
        pesos_iniciales = np.array([0, 1, 2])
        bias = 1

        # Primera capa oculta
        self.h1_1 = Neurona(pesos_iniciales, bias)
        self.h1_2 = Neurona(pesos_iniciales, bias)

        # Segunda capa oculta
        self.h2_1 = Neurona(np.array([0, 1]), bias)
        self.h2_2 = Neurona(np.array([0, 1]), bias)

        # Capa de salida
        self.o1 = Neurona(np.array([0, 1]), bias)

    def alimentacion(self, x):
        # Primera capa oculta
        out_h1_1 = self.h1_1.activacion(x)
        out_h1_2 = self.h1_2.activacion(x)

        # Segunda capa oculta
        out_h2_1 = self.h2_1.activacion(np.array([out_h1_1, out_h1_2]))
        out_h2_2 = self.h2_2.activacion(np.array([out_h1_1, out_h1_2]))

        # Capa de salida
        out_o1 = self.o1.activacion(np.array([out_h2_1, out_h2_2]))

        return out_o1

# Crear la red neuronal y alimentarla con los valores [10, 30, 40]
network = MiRedNeuronal()
x = np.array([10, 30, 40])
resultado = network.alimentacion(x)
resultado


0.8677026536525567

La red neuronal con los parámetros y estructura especificados produce una salida de aproximadamente 0.8677 cuando se alimenta con los valores de entrada [10,30,40].

Para analizar los resultados, primero, cabe mencionar que todos los pesos y sesgos están inicializados con valores constantes, lo que significa que no hay ningún proceso de entrenamiento involucrado aquí. Estos valores son simplemente una inicialización y, en un escenario real, se ajustarían mediante un proceso de entrenamiento para que la red pueda realizar tareas específicas.

La función sigmoide, que se utiliza como función de activación, mapea cualquier valor de entrada a un número entre 0 y 1. Dado que todos los pesos y sesgos son constantes en esta red, la salida será la misma para cualquier entrada, excepto que se escala según el valor de la entrada.

**Ponte a prueba 2**

Una empresa de fabricación de relojes inteligentes desea clasificar sus relojes en diferentes rangos de precios según sus características. La empresa ha recopilado datos de varios modelos de relojes inteligentes y desea utilizar estos datos para predecir el rango de precios de los nuevos modelos que producirán en el futuro.

**Variables**:
1. `battery_duration`: Duración de la batería en horas (continuo).
2. `screen_size`: Tamaño de la pantalla en pulgadas (continuo).
3. `memory`: Memoria interna en GB (continuo).
4. `has_GPS`: Si el reloj tiene GPS (0: No, 1: Sí).
5. `has_Bluetooth`: Si el reloj tiene Bluetooth (0: No, 1: Sí).
6. `water_resistant`: Si el reloj es resistente al agua (0: No, 1: Sí).
7. `price_range`: Rango de precios del reloj (0: Bajo, 1: Medio, 2: Alto, 3: Premium).

Puedes encontrar el conjunto de datos en el siguiente enlace: https://drive.google.com/file/d/1yvZnllbO5wRdPgj1P_zDUNePoKz0i4wJ/view?usp=sharing

Basado en este conjunto de datos, deberás:

1. Dividir los datos en conjuntos de entrenamiento y prueba.
2. Normalizar o estandarizar las características.
3. Construir un modelo de red neuronal usando Keras (o cualquier otro marco que prefieras).
4. Entrenar el modelo con el conjunto de datos de entrenamiento.
5. Evaluar el modelo con el conjunto de datos de prueba.
6. Hacer predicciones con el modelo y analizar los resultados.


In [ ]:
#Respuesta con comentarios para explicar el proceso

In [ ]:
### Importamos las librerías necesarias
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# 1. Cargar el conjunto de datos
# Reemplaza por la ubicación de tu conjunto de datos
data = pd.read_csv("/content/drive/MyDrive/Bootcamp Data Science/Datasets/smartwatch_data.csv")

# 2. Dividir los datos en conjuntos de entrenamiento y prueba
X = data.drop("price_range", axis=1).values
y = data["price_range"].values
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 3. Normalizar las características
scaler = StandardScaler()
X_train_normalized = scaler.fit_transform(X_train)
X_test_normalized = scaler.transform(X_test)

# 4. Construir el modelo de red neuronal
model = Sequential([
    Dense(128, activation='relu', input_shape=(X_train_normalized.shape[1],)),
    Dense(64, activation='relu'),
    Dense(32, activation='relu'),
    Dense(4, activation='softmax')
])

# 5. Compilar el modelo
model.compile(optimizer=Adam(), loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# 6. Entrenar el modelo
model.fit(X_train_normalized, y_train, epochs=20, batch_size=32, verbose=1)

# 7. Evaluar el modelo
loss, accuracy = model.evaluate(X_test_normalized, y_test, verbose=0)
print(f"Accuracy: {accuracy * 100:.2f}%")

Epoch 1/20
25/25 [==============================] - 1s 3ms/step - loss: 1.3003 - accuracy: 0.4638
Epoch 2/20
25/25 [==============================] - 0s 2ms/step - loss: 1.0426 - accuracy: 0.5962
Epoch 3/20
25/25 [==============================] - 0s 2ms/step - loss: 0.8244 - accuracy: 0.6413
Epoch 4/20
25/25 [==============================] - 0s 2ms/step - loss: 0.7094 - accuracy: 0.6712
Epoch 5/20
25/25 [==============================] - 0s 2ms/step - loss: 0.6094 - accuracy: 0.7437
Epoch 6/20
25/25 [==============================] - 0s 3ms/step - loss: 0.5027 - accuracy: 0.8213
Epoch 7/20
25/25 [==============================] - 0s 3ms/step - loss: 0.4152 - accuracy: 0.8700
Epoch 8/20
25/25 [==============================] - 0s 2ms/step - loss: 0.3498 - accuracy: 0.8975
Epoch 9/20
25/25 [==============================] - 0s 2ms/step - loss: 0.2946 - accuracy: 0.9175
Epoch 10/20
25/25 [==============================] - 0s 3ms/step - loss: 0.2577 - accuracy: 0.9312
Epoch 11/20
25/25 [

In [ ]:
# 8. Generar 20 predicciones
predictions = model.predict(X_test_normalized[:20])
predicted_classes = np.argmax(predictions, axis=1)

print("20 Predicciones:", predicted_classes)
print("Etiquetas Reales:", y_test[:20])

1/1 [==============================] - 0s 143ms/step
20 Predicciones: [1 3 2 2 1 1 1 2 0 1 3 2 3 1 1 0 2 0 0 1]
Etiquetas Reales: [1 3 2 2 1 1 1 3 0 1 3 2 3 1 1 0 2 0 0 1]


Comparando ambas listas, podemos ver que el modelo ha acertado en todas las predicciones para estas 20 muestras. Esto indica que el modelo está haciendo un trabajo excepcional en la clasificación de los rangos de precios de los relojes inteligentes basándose en las características proporcionadas.

Dado que el conjunto de datos es ficticio y fue generado, el modelo pudo aprender las relaciones subyacentes con bastante facilidad, resultando en una alta precisión. Sin embargo, en conjuntos de datos del mundo real, es posible que no veamos un rendimiento tan alto y es posible que tengamos que realizar más ajustes y refinamientos en el modelo.

**Reto semanal**

Usa el siguiente dataset de precios de dispositivos móviles: https://www.kaggle.com/datasets/iabhishekofficial/mobile-price-classification

Seguramente te parecerá familiar porque ya lo hemos usado antes.

Haz lo siguiente:
- Genera un modelo de entrenamiento con Keras.
- Experimenta con la cantidad de nodos y capas.
- Experimenta con diferentes optimizadores al compilar el modelo.
- Entrena el modelo y obtén la precisión.
- Genera 20 predicciones y analiza los resultados.

In [ ]:
#Respuesta con comentarios para explicar el proceso

In [ ]:
## Importamos las librerías necesarias
!pip install tensorflow
import pandas as pd
import tensorflow as tf
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam, SGD, RMSprop

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# Cargar los conjuntos de datos de entrenamiento y prueba
# Reemplaza por la ubicación de tu conjunto de datos
train_data = pd.read_csv("/content/drive/MyDrive/Bootcamp Data Science/Datasets/train.csv")

X = train_data.drop("price_range", axis=1).values
y = train_data["price_range"].values

# Normalizar las características
scaler = StandardScaler()
X_normalized = scaler.fit_transform(X)

# Dividir el conjunto de datos en entrenamiento y validación
X_train, X_val, y_train, y_val = train_test_split(X_normalized, y, test_size=0.2, random_state=42)

# Mostrar las primeras filas del conjunto de datos de entrenamiento
train_data.head()

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  m_dep  \
0            842     0          2.2         0   1       0           7    0.6   
1           1021     1          0.5         1   0       1          53    0.7   
2            563     1          0.5         1   2       1          41    0.9   
3            615     1          2.5         0   0       0          10    0.8   
4           1821     1          1.2         0  13       1          44    0.6   

   mobile_wt  n_cores  ...  px_height  px_width   ram  sc_h  sc_w  talk_time  \
0        188        2  ...         20       756  2549     9     7         19   
1        136        3  ...        905      1988  2631    17     3          7   
2        145        5  ...       1263      1716  2603    11     2          9   
3        131        6  ...       1216      1786  2769    16     8         11   
4        141        2  ...       1208      1212  1411     8     2         15   

   three_g  touch_screen  wifi  price_range  
0        0             0     1            1  
1        1             1     0            2  
2        1             1     0            2  
3        1             0     0            2  
4        1             1     0            1  

[5 rows x 21 columns]

El conjunto de datos de entrenamiento ha sido cargado correctamente. Contiene varias características como battery_power, blue (Bluetooth), clock_speed, entre otras, y la variable objetivo price_range que indica el rango de precio del dispositivo móvil.

El siguiente paso es construir el modelo de red neuronal con Keras.

In [ ]:
# Normalizar las características
scaler = StandardScaler()
X_train_normalized = scaler.fit_transform(X_train)
X_test_normalized = scaler.transform(X_val)

# Construcción del modelo
model = Sequential([
    Dense(128, activation='relu', input_shape=(X_train_normalized.shape[1],)),
    Dense(64, activation='relu'),
    Dense(32, activation='relu'),
    Dense(4, activation='softmax')  # Hay 4 clases en price_range: 0, 1, 2, y 3
])

# Compilar el modelo usando el optimizador Adam como inicial
model.compile(optimizer=Adam(), loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# Resumen del modelo
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_4 (Dense)             (None, 128)               2688      
                                                                 
 dense_5 (Dense)             (None, 64)                8256      
                                                                 
 dense_6 (Dense)             (None, 32)                2080      
                                                                 
 dense_7 (Dense)             (None, 4)                 132       
                                                                 
Total params: 13156 (51.39 KB)
Trainable params: 13156 (51.39 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


Se observa la configuración inicial de la red, con una estructura básica.

**Entrenamiento y precisión del modelo**

In [ ]:
# Entrenamiento del modelo
history = model.fit(X_train_normalized, y_train, validation_data=(X_test_normalized, y_val), epochs=20, batch_size=32, verbose=1)

# Evaluación del modelo
loss, accuracy = model.evaluate(X_val, y_val, verbose=0)
print(f"Accuracy: {accuracy * 100:.2f}%")

Epoch 1/20
50/50 [==============================] - 2s 10ms/step - loss: 1.2351 - accuracy: 0.4769 - val_loss: 0.9446 - val_accuracy: 0.6450
Epoch 2/20
50/50 [==============================] - 0s 3ms/step - loss: 0.6801 - accuracy: 0.7725 - val_loss: 0.4405 - val_accuracy: 0.8475
Epoch 3/20
50/50 [==============================] - 0s 3ms/step - loss: 0.3307 - accuracy: 0.9000 - val_loss: 0.2939 - val_accuracy: 0.8950
Epoch 4/20
50/50 [==============================] - 0s 4ms/step - loss: 0.1993 - accuracy: 0.9419 - val_loss: 0.2256 - val_accuracy: 0.9125
Epoch 5/20
50/50 [==============================] - 0s 3ms/step - loss: 0.1523 - accuracy: 0.9488 - val_loss: 0.2136 - val_accuracy: 0.9125
Epoch 6/20
50/50 [==============================] - 0s 7ms/step - loss: 0.1092 - accuracy: 0.9706 - val_loss: 0.1890 - val_accuracy: 0.9300
Epoch 7/20
50/50 [==============================] - 0s 6ms/step - loss: 0.0926 - accuracy: 0.9744 - val_loss: 0.1904 - val_accuracy: 0.9100
Epoch 8/20
50/50 [=

Se observa que el accuracy de la red entrenada es del 92.5%. Podríamos probar distintos optimizadores y configuraciones de la red buscando incrementar el desempeño.

In [ ]:
# Generar 20 predicciones
predictions = model.predict(X_val[:20])
predicted_classes = np.argmax(predictions, axis=1)

print("20 Predicciones:", predicted_classes)
print("Etiquetas Reales:", y_val[:20])

1/1 [==============================] - 0s 182ms/step
20 Predicciones: [0 2 1 3 1 1 2 0 3 1 0 1 2 3 3 3 3 3 1 0]
Etiquetas Reales: [0 2 1 3 1 1 2 0 3 1 0 0 2 3 3 2 3 3 1 0]


Comparando ambas listas, podemos ver que el modelo cometió 2 errores en las 20 predicciones.

Esto significa que el modelo tuvo un 90% de precisión en estas 20 predicciones, lo cual es bastante bueno. Sin embargo, es importante tener en cuenta que este es solo un subconjunto de los datos de prueba, y la precisión general del modelo puede variar cuando se evalúa en todo el conjunto de datos de prueba.

Los resultados sugieren que el modelo es bastante efectivo en la clasificación de los rangos de precios de los dispositivos móviles, pero siempre hay margen de mejora. Podemos experimentar con diferentes arquitecturas de red, optimizadores, técnicas de regularización o incluso aumentar el número de épocas para ver si es posible mejorar aún más el rendimiento del modelo.